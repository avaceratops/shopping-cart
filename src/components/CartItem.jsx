import { formatPrice } from '../utils/formatting';
import QuantitySelector from './QuantitySelector';

export default function CartItem({ product, quantity, updateCartItem, removeFromCart }) {
  const handleRemoveClick = () => {
    removeFromCart(product.id);
  };

  const handleQuantityChange = (newValue) => {
    updateCartItem(product.id, newValue);
  };

  return (
    <li key={product.id} className="flex py-6">
      <div className="h-24 w-24 flex-shrink-0 overflow-hidden rounded border">
        <img className="h-full w-full object-cover object-center" src={product.image} alt="" />
      </div>

      <div className="ml-4 flex flex-1 flex-col">
        <div>
          <div className="flex justify-between text-base font-medium text-gray-900">
            <h3>{product.name}</h3>
            <p className="ml-4">{formatPrice(product.price)}</p>
          </div>

          <p className="mt-1 text-sm text-gray-500">{product.faction}</p>
        </div>

        <div className="flex flex-1 items-end justify-between text-sm">
          <QuantitySelector
            id={product.id}
            initialValue={quantity}
            maxValue={product.stock}
            onChange={handleQuantityChange}
            isCart={true}
          />

          <div className="flex">
            <button
              type="button"
              className="font-medium text-indigo-600 hover:text-indigo-500"
              onClick={handleRemoveClick}
            >
              Remove
            </button>
          </div>
        </div>
      </div>
    </li>
  );
}
