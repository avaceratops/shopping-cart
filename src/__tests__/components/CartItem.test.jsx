import { vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import CartItem from '../../components/CartItem';

const product = {
  id: '00001',
  name: 'Norn Emissary',
  price: 7000,
  faction: 'Tyranids',
  image:
    '/images/warhammer-40k/https___trade.games-workshop.com_assets_2023_08_99120106064_NornAssimilatorStock.jpg',
  sku: 99120106064,
  stock: 10,
  sold: 48,
};

it('renders cart item correctly', () => {
  const { container } = render(<CartItem product={product} />);
  expect(container).toMatchSnapshot();
});

it('calls removeFromCart correctly when "Remove" button is clicked', async () => {
  const removeFromCart = vi.fn();
  const user = userEvent.setup();
  render(<CartItem product={product} removeFromCart={removeFromCart} />);

  const removeButton = screen.getByRole('button', { name: /remove/i });
  await user.click(removeButton);

  expect(removeFromCart).toHaveBeenCalledWith('00001');
});
