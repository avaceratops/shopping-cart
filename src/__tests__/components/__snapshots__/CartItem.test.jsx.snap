// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders cart item correctly 1`] = `
<div>
  <li
    class="flex py-6"
  >
    <div
      class="h-24 w-24 flex-shrink-0 overflow-hidden rounded border"
    >
      <img
        alt=""
        class="h-full w-full object-cover object-center"
        src="/images/warhammer-40k/https___trade.games-workshop.com_assets_2023_08_99120106064_NornAssimilatorStock.jpg"
      />
    </div>
    <div
      class="ml-4 flex flex-1 flex-col"
    >
      <div>
        <div
          class="flex justify-between text-base font-medium text-gray-900"
        >
          <h3>
            Norn Emissary
          </h3>
          <p
            class="ml-4"
          >
            Â£70.00
          </p>
        </div>
        <p
          class="mt-1 text-sm text-gray-500"
        >
          Tyranids
        </p>
      </div>
      <div
        class="flex flex-1 items-end justify-between text-sm"
      >
        <article>
          <label
            class="sr-only"
            for="cart-00001"
          >
            Quantity
          </label>
          <select
            class="_select_31f26a"
            id="cart-00001"
            name="cart-00001"
          >
            <option
              value="1"
            >
              1
            </option>
            <option
              value="2"
            >
              2
            </option>
            <option
              value="3"
            >
              3
            </option>
            <option
              value="4"
            >
              4
            </option>
            <option
              value="5"
            >
              5
            </option>
            <option
              value="6"
            >
              6
            </option>
            <option
              value="7"
            >
              7
            </option>
            <option
              value="8"
            >
              8
            </option>
          </select>
        </article>
        <div
          class="flex"
        >
          <button
            class="font-medium text-indigo-600 hover:text-indigo-500"
            type="button"
          >
            Remove
          </button>
        </div>
      </div>
    </div>
  </li>
</div>
`;
